Orange/widgets/data/owtable.py

```diff
@@ -69,24 +74,157 @@ def __init__(self, sourcedata, parent=None):
+                elif view_col < self.Y_dim + self.M_dim:
+                    self.cache[(view_row, view_col)] = M[chunk_row][chunk_col - Y_inview]
+                else:
+                    self.cache[(view_row, view_col)] = X[chunk_row][chunk_col - Y_inview - M_inview]
+
+    def request_data(self):
+        cache_interval = 0.02
+        elapsed = perf_counter() - self.last_cache_time
+        self.timer.stop()
+        self.timer.start(max(1, int((cache_interval - elapsed) * 1000)))
+
+    def on_timeout(self):
+        self.last_cache_time = perf_counter()
+        self.timer.stop()
+        self.update_viewport()
+        self.start(self.update_cache)
+
+    def on_done(self, _result):
+        self.dataChanged.emit(self.index(self.viewport.top, self.viewport.left),
+                              self.index(self.viewport.bottom, self.viewport.right))
+
+    def sort(self, *args, **kwargs):
+        super().sort(*args, **kwargs)
+        if isinstance(self.source, DaskTable):
+            # upon sorting cache must be reset completely
+            self.cache.clear()
+            self.update_viewport()
+            self.start(self.update_cache)
+
+    def get_value(self, row, col):
+        if isinstance(self.source, DaskTable):
+            try:
+                return self.cache[(row, col)]
+            except KeyError:
+                self.request_data()
+                return
+        elif self.X_density == Storage.DENSE:
+            row = self.mapToSourceRows(row)
+            if col < self.Y_dim:
+                if self.source.Y.ndim == 1:
+                    return self.source.Y[row]
+                return self.source.Y[row, col]
+            elif self.Y_dim <= col < self.Y_dim + self.M_dim:
+                return self.source.metas[row, col - self.Y_dim]
+            else:
+                return self.source.X[row, col - self.Y_dim - self.M_dim]
+
     def data(self, index, role=Qt.DisplayRole,
              # for faster local lookup
              _BarRole=gui.TableBarItem.BarRole):
         # pylint: disable=arguments-differ
-        if role == _BarRole and self._continuous[index.column()]:
-            val = super().data(index, TableModel.ValueRole)
-            if val is None or isnan(val):
-                return None
+        if self.X_density != Storage.DENSE:
+            return super().data(index, role)
 
-            dist = super().data(index, TableModel.VariableStatsRole)
+        row = index.row()
+        col = 0 if role is self.ClassValueRole else index.column()
+
+        value = self.get_value(row, col)
+        if value is None:
+            return
+        coldesc = self.columns[col]
+
+        if role == Qt.DisplayRole:
+            return coldesc.var.str_val(value)
+        elif role in (self.ValueRole, Qt.EditRole) and isinstance(coldesc, self.Column):
+            return Value(coldesc.var, value)
+        elif role == Qt.BackgroundRole:
```

Comment:
I understand you need to re-implement this function for DisplayRole and ValueRole and all roles defined in the previous version, but couldn't you call the super, but couldn't you call the superclass for the others?

Or are the other reasons why you put them here? Perhaps because you'd like to move this code into TableModel?

Reply:
As soon as I call super().data() it will read the entire row though. So I need to avoid it entirely and rewrite all of the code or find a clean way to implement it into the base class (with tests and all).